docs: Implement cascading deletion for obsolete TrackedEntities[DHIS2-15066] (#196)
dhis2/dhis2-releases: releases/2.42/migration-notes.md

  - [Inconsistent data](#inconsistent-data)
    - [Tracker](#tracker)

> 	   - [Null Organisation Unit](#null-organisation-unit)
>        - [Null Tracked Entity Type](#null-tracked-entity-type)

Reviewer: ```suggestion
       - [Null Organisation Unit](#null-organisation-unit)
       - [Null Tracked Entity Type](#null-tracked-entity-type)
```

and en.uid = '{ENROLLMENT_UID}'
AND en.programid in (select programid from program where type = 'WITH_REGISTRATION');
```

### Null Tracked Entity Type

The `TrackedEntity`, previously known as `TrackedEntityInstance`, is required to have a specific type, such as person, place, equipment, or area. However, this constraint was not enforced at the database level, leading to inconsistent data. To ensure data integrity moving forward, we need to enforce this requirement by making the `trackedentitytypeid` column in the `trackedentity` table non-nullable.

#### Checking for Null Values

To check for any NULL values in this column, you can use the following SQL script. If it returns a value greater than 0, it indicates the presence of inconsistent data in the system.

##### For 2.41 Instances:

```sql
SELECT COUNT(1)
FROM trackedentity
WHERE trackedentitytypeid IS NULL;
```
##### For <= 2.40 Instances:

```sql
SELECT COUNT(1)
FROM trackedentityinstance
WHERE trackedentitytypeid IS NULL;
```

> #### Fixing Null Values

Reviewer: As we are doing for null orgUnit, we should provide a way for the user to fix the data and put it as a first option.
Deleting data should be the last resort here.
What we did in the other section was to suggest to find a similar entity and to transfer the metadata from that entity to the inconsistent entity.

 
END;
$$;
```

### Null Tracked Entity Type

The `TrackedEntity`, previously known as `TrackedEntityInstance`, is required to have a specific type, such as person, place, equipment, or area. However, this constraint was not enforced at the database level, leading to inconsistent data. To ensure data integrity moving forward, we need to enforce this requirement by making the `trackedentitytypeid` column in the `trackedentity` table non-nullable.

#### Checking for Null Values

To check for any NULL values in this column, you can use the following SQL script. If it returns a value greater than 0, it indicates the presence of inconsistent data in the system.

##### For 2.41 Instances:

> ```sql

Reviewer: Here we shouldn't show entities that we can fix in the migration, so we should filter out any tracked entity that has an enrollment with a program that has a defined trackedEntityType

Reviewer: I still think we should change this query to not show the entities that will be fixed by the migration

SELECT COUNT(1)
FROM trackedentity
WHERE trackedentitytypeid IS NULL;
```
##### For <= 2.40 Instances:

> ```sql

Reviewer: Same here

SELECT COUNT(1)
FROM trackedentityinstance
WHERE trackedentitytypeid IS NULL;
```

#### Fixing Null Values

Starting from version v42, NULL values are no longer allowed in the trackedentitytypeid column. The migration attempted to address the invalid data, but it was unsuccessful. There are two options going forward.
- Change the `NULL` value to a valid trackedentitytypeid. ([Assign trackedentitytyeid to tracked entity](#assign-tracked-entity-type))
- Completely remove invalid trackedentity record. ([Delete trackedentities](#deleting-invalid-trackedenetities))

##### Assign tracked entity type

> To assign valid trackedentitytypeid. It can be fetched from program via enrollment.

Reviewer: This is what we are doing in the migration, we shouldn't let them doing here manually.
This section should be about the tracked entities that the migration wasn't able to fix, so we should provide a way, as naive as possible, to make them manually fix the data.
In the null organisation unit what I did was to make them find a similar event/enrollment and copy the organisation unit from there.

Here, as there shouldn't be so many tracked entity types in the system, we could provide a SQL snippet to let them assign one tracked entity type of their choose.

```sql
UPDATE trackedentity SET trackedentitytypeid = (
SELECT program.trackedentitytypeid FROM program
        JOIN enrollment ON program.programid = enrollment.programid
        WHERE enrollment.trackedentityid = trackedentity.trackedentityid)
WHERE trackedentitytypeid IS NULL AND EXISTS (
SELECT 1 FROM enrollment JOIN program ON enrollment.programid = program.programid
        WHERE enrollment.trackedentityid = trackedentity.trackedentityid
);
```

##### Deleting invalid trackedenetities

> The following script can be used to remove all trackedentities with null in `trackedentitytypeid` column.

Reviewer: Here we need to be really careful and make them notice that this SQL will remove all the tracked entities, even the ones that could potentially be fixed.

Maybe should we provide an alternative one that only remove the not automatically fixable tracked entity?

WHERE trackedentitytypeid IS NULL;
```

#### Fixing Null Values

Starting from version v42, NULL values are no longer allowed in the trackedentitytypeid column. The migration attempted to address the invalid data, but it was unsuccessful. There are two options going forward.
- Change the `NULL` value to a valid trackedentitytypeid. ([Assign trackedentitytyeid to tracked entity](#assign-tracked-entity-type))
- Completely remove invalid trackedentity record. ([Delete trackedentities](#deleting-invalid-trackedenetities))

##### Assign tracked entity type
To assign valid trackedentitytypeid 
- Use the following command to list all the trackedentitytypes currently available in the system. This query retrieves the uid and name of all trackedentitytypes in your database. Review the output to identify the most appropriate trackedentitytype.

```sql
SELECT uid, name FROM trackedentitytype;
```

- Replace {REFERENCE_UID} with the uid of the selected trackedentitytype and execute the below mentioned command.

```sql
UPDATE trackedentity SET trackedentitytypeid=( SELECT trackedentitytypeid FROM trackedentitytype WHERE uid='{REFERENCE_UID}') WHERE trackedentitytypeid IS NULL;
```

> ##### Deleting invalid trackedenetities

Reviewer: ```suggestion
##### Deleting invalid tracked entities
```

The script below can be used to remove all trackedentity records where the trackedentitytypeid column is null. Use this script with caution, as it may result in permanent data loss. 

> ```plsql

Reviewer: This script should remove only the entities that are not going to be fixed by the migration

        WHERE enrollmentid IN (SELECT enrollmentid FROM enrollment_ids);

        DELETE FROM event
        WHERE enrollmentid IN (SELECT enrollmentid FROM enrollment_ids);

        DELETE FROM programmessage
        WHERE trackedentityid IN (SELECT trackedentityid FROM te);

        DELETE FROM relationshipitem
        WHERE trackedentityid IN (SELECT trackedentityid FROM te);

        DELETE FROM trackedentityattributevalue
        WHERE trackedentityid IN (SELECT trackedentityid FROM te);

        DELETE FROM trackedentityattributevalueaudit
        WHERE trackedentityid IN (SELECT trackedentityid FROM te);
      
        DELETE FROM trackedentityprogramowner
        WHERE trackedentityid IN (SELECT trackedentityid FROM te);

        DELETE FROM programtempowner
        WHERE trackedentityid IN (SELECT trackedentityid FROM te);

        DELETE FROM programtempownershipaudit
        WHERE trackedentityid IN (SELECT trackedentityid FROM te);

        DELETE FROM programownershiphistory
        WHERE trackedentityid IN (SELECT trackedentityid FROM te);

        DELETE FROM enrollment
        WHERE trackedentityid IN (SELECT trackedentityid FROM te);
		

>         WITH deleted AS (DELETE FROM trackedentity WHERE trackedentitytypeid IS NULL RETURNING *) SELECT COUNT(*) INTO deleted_count FROM deleted;

Reviewer: Here we should delete only the invalid tracked entities. Without the NOT EXISTS clause, also the fixable ones are going to be deleted

```suggestion
        WITH deleted AS (DELETE FROM trackedentity WHERE trackedentitytypeid IS NULL AND NOT EXISTS (
           SELECT 1
            FROM enrollment e JOIN program p on e.programid = p.programid
            WHERE e.trackedentityid = te.trackedentityid and p.trackedentitytypeid IS NOT NULL
        ) 
        RETURNING *) SELECT COUNT(*) INTO deleted_count FROM deleted;
```

 
END;
$$;
```

### Null Tracked Entity Type

The `TrackedEntity`, previously known as `TrackedEntityInstance`, is required to have a specific type, such as person, place, equipment, or area. However, this constraint was not enforced at the database level, leading to inconsistent data. To ensure data integrity moving forward, we need to enforce this requirement by making the `trackedentitytypeid` column in the `trackedentity` table non-nullable.

#### Checking for Null Values

To check for any NULL values in this column, you can use the following SQL script. If it returns a value greater than 0, it indicates the presence of inconsistent data which migration was not able to fix.

##### For 2.41 Instances:

```sql

> SELECT COUNT(1)
> FROM trackedentity te
> WHERE te.trackedentitytypeid IS NULL
>   AND NOT EXISTS (
>     SELECT 1
>     FROM enrollment e
>     WHERE e.trackedentityid = te.trackedentityid
>   );

Reviewer: ```suggestion
SELECT COUNT(1)
FROM trackedentity te
WHERE te.trackedentitytypeid IS NULL
  AND NOT EXISTS (
    SELECT 1
    FROM enrollment e JOIN program p on e.programid = p.programid
    WHERE e.trackedentityid = te.trackedentityid and p.trackedentitytypeid IS NOT NULL
);
```

```
##### For <= 2.40 Instances:

```sql

> SELECT COUNT(1)
> FROM trackedentityinstance te
> WHERE te.trackedentitytypeid IS NULL
>   AND NOT EXISTS (
>     SELECT 1
>     FROM programinstance pi
>     WHERE pi.trackedentityinstanceid = te.trackedentityinstanceid
>   );

Reviewer: ```suggestion
SELECT COUNT(1)
FROM trackedentityinstance te
WHERE te.trackedentitytypeid IS NULL
  AND NOT EXISTS (
    SELECT 1
    FROM programinstance pi JOIN program p ON pi.programid = p.programid
    WHERE pi.trackedentityinstanceid = te.trackedentityinstanceid and p.trackedentitytypeid IS NOT NULL
  );
```

```

#### Fixing Null Values

Starting from version v42, NULL values are no longer allowed in the trackedentitytypeid column. The migration attempted to address the invalid data, but it was unsuccessful. There are two options going forward.
- Change the `NULL` value to a valid trackedentitytypeid. ([Assign trackedentitytyeid to tracked entity](#assign-tracked-entity-type))

> - Completely remove invalid trackedentity record. ([Delete trackedentities](#deleting-invalid-trackedenetities))

Reviewer: ```suggestion
- Completely remove invalid trackedentity record. ([Delete trackedentities](#deleting-invalid-tracked-entities))
```

##### Assign tracked entity type
To assign valid trackedentitytypeid 
- Use the following command to list all the trackedentitytypes currently available in the system. This query retrieves the uid and name of all trackedentitytypes in your database. Review the output to identify the most appropriate trackedentitytype.

```sql
SELECT uid, name FROM trackedentitytype;
```

- Replace {REFERENCE_UID} with the uid of the selected trackedentitytype and execute the below mentioned command.

```sql

> UPDATE trackedentity SET trackedentitytypeid=( SELECT trackedentitytypeid FROM trackedentitytype WHERE uid='{REFERENCE_UID}') WHERE trackedentitytypeid IS NULL;

Reviewer: This query is changing all the trackedEntities with `null` tracked entity type at once.
Can we change the count queries to return the tracked entity uids to be fixed instead of just a count?
And then add them here as a parameter, like ` AND trackedentityid IN ({uids})`?

```

> ##### Deleting invalid tracked enetities

Reviewer: ```suggestion
##### Deleting invalid tracked entities
```

    deleted_count INT := 0;  -- Variable to keep track of deleted TrackedEntity count
BEGIN
    CREATE TEMP TABLE te AS (
        SELECT trackedentityid
        FROM trackedentity
        WHERE trackedentitytypeid IS NULL
    );

    CREATE TEMP TABLE enrollment_ids AS (
        SELECT enrollmentid
        FROM enrollment
        WHERE trackedentityid IN (SELECT trackedentityid FROM te)
    );

    CREATE TEMP TABLE event_ids AS (
        SELECT eventid
        FROM event
        WHERE enrollmentid IN (SELECT enrollmentid FROM enrollment_ids)
    );

    CREATE TEMP TABLE te_pm AS (
        SELECT id
        FROM programmessage
        WHERE trackedentityid IN (SELECT trackedentityid FROM te)
    );

    CREATE TEMP TABLE pi_pm AS (
        SELECT id
        FROM programmessage
        WHERE enrollmentid IN (SELECT enrollmentid FROM enrollment_ids)
    );

    CREATE TEMP TABLE event_pm AS (
        SELECT id
        FROM programmessage
        WHERE eventid IN (SELECT eventid FROM event_ids)
    );

    SELECT COUNT(trackedentityid)
    INTO invalid_count
    FROM trackedentity
    WHERE trackedentitytypeid IS NULL AND NOT EXISTS (

>             SELECT 1
>             FROM enrollment
>             WHERE enrollment.trackedentityid = trackedentity.trackedentityid

Reviewer: ```suggestion
            SELECT 1
            FROM enrollment e JOIN program p on e.programid = p.programid
            WHERE e.trackedentityid = te.trackedentityid and p.trackedentitytypeid IS NOT NULL
```

Author: fixed
